lib/change_stream.js

```diff
@@ -534,8 +497,108 @@ function processNewChange(args) {
+  return callback(undefined, change);
+}
+
+function processError(changeStream, error, callback) {
+  const topology = changeStream.topology;
+  const cursor = changeStream.cursor;
+
+  // If the change stream has been closed explictly, do not process error.
+  if (changeStream.isClosed()) {
+    if (callback) callback(new MongoError('ChangeStream is closed'));
+    return;
+  }
+
+  // if the resume succeeds, continue with the new cursor
+  function resumeWithCursor(newCursor) {
+    changeStream.cursor = newCursor;
+    processResumeQueue(changeStream);
+  }
+
+  // otherwise, raise an error and close the change stream
+  function unresumableError(err) {
+    if (!callback) {
+      changeStream.emit('error', err);
+      changeStream.emit('close');
+    }
+    processResumeQueue(changeStream, err);
+    changeStream.closed = true;
+  }
+
+  if (cursor && isResumableError(error, maxWireVersion(cursor.server))) {
+    changeStream.cursor = undefined;
+
+    // stop listening to all events from old cursor
+    ['data', 'close', 'end', 'error'].forEach(event => cursor.removeAllListeners(event));
+
+    // close internal cursor, ignore errors
+    cursor.close();
+
+    waitForTopologyConnected(topology, { readPreference: cursor.options.readPreference }, err => {
+      // if the topology can't reconnect, close the stream
+      if (err) return unresumableError(err);
+
+      // create a new cursor, preserving the old cursor's options
+      const newCursor = createChangeStreamCursor(changeStream, cursor.resumeOptions);
+
+      // attempt to continue in emitter mode
+      if (!callback) return resumeWithCursor(newCursor);
+
+      // attempt to continue in iterator mode
+      newCursor.hasNext(err => {
+        // if there's an error immediately after resuming, close the stream
+        if (err) return unresumableError(err);
+        resumeWithCursor(newCursor);
+      });
+    });
+    return;
+  }
+
+  if (!callback) return changeStream.emit('error', error);
+  return callback(error);
+}
+
+/**
+ * Safely provides a cursor across resume attempts
+ *
+ * @param {ChangeStream} changeStream the parent ChangeStream
+ * @param {function} callback gets the cursor or error
+ * @param {ChangeStreamCursor} [oldCursor] when resuming from an error, carry over options from previous cursor
+ */
+function getCursor(changeStream, callback) {
+  if (changeStream.isClosed(true)) {
+    callback(new MongoError('ChangeStream is closed.'));
+    return;
+  }
+
+  // if a cursor exists and it is open, return it
+  if (changeStream.cursor) {
+    callback(undefined, changeStream.cursor);
+    return;
+  }
+
+  // no cursor, queue callback until topology reconnects
+  changeStream[kResumeQueue].push(callback);
+}
+
+/**
+ * Drain the resume queue when a new has become available
+ *
+ * @param {ChangeStream} changeStream the parent ChangeStream
+ * @param {?ChangeStreamCursor} changeStream.cursor the new cursor
```

Comment:
nit: for plain jsdoc I think the preference would be to write this as:
```
@param {ChangeStreamCursor} [changeStream.cursor]
````
but also I think both TypeScript and JavaScript prefer the question mark at the end of the symbol, rather than before.

Reply:
This is something that's not entirely clear to me; is there a difference between an optional and a ["nullable type"](https://jsdoc.app/tags-type.html)?

I.e. I can write a function two ways. 
```
function optionalMiddleParam(a, b, c) {
  // handle options
  console.log({ a, b, c });
}
function nullableMiddleParam(a, b, c) {
  // handle options
  console.log({ a, b, c });
}
```

`optionalMiddleParam(1, 2)`  => outputs `{ a: 1, b: undefined, c: 2}`
`nullableMiddleParam(1, 2)` => outputs `{ a: 1, b: 2, c: undefined }`
and instead needs to be called as
`nullableMiddleParam(1, undefined, 2)` or `nullableMiddleParam(1, null, 2)`

My thinking was to use `[param]` vs `{?type}` to distinguish the two, but I wasn't really able to get a definitive answer from reading the jsdoc and typescript docs about nullable types vs optional params.